### PR TITLE
Create application that handles dev account creation

### LINKED ISSUE #422: Create application that handles dev account creation
Create a program that runs in the terminal that automatically creates an account for development purposes.

When we run it, the terminal should look like:

```
Welcome to SCE's dev account creation!
Enter email name:
Enter password:
Enter account priviledge: 
One moment...
Done! You may now sign in with <email they entered> with password <password they entered>.
```

### PR DESCRIPTION
**Pull request by hsorathia**
Resolves #422

How to test this PR out:
1. Clone / checkout the `create-user` branch
2. Run `npm install` in the terminal
3. Start mongodb
4. Run the creation script with the command `npm run create-user`. You can try variations with this like running the server/frontend and also running this script at the same time.
5. Run the server/frontend (if you haven't already) to test the account you just created

~~Everytime you run the script you have to discard the terminal afterwards.~~ You no longer have to discard the terminal everytime you run the script.

Bad email example(s):
image

image

No password & password provided sample executions:
image

image

### CODE REVIEW

File: api/util/registerUser.js [hunk 1]
```diff
@@ -24,7 +24,10 @@ async function registerUser(userToAdd){
       email: userToAdd.email.toLowerCase(),
       major: userToAdd.major || ''
     });
+<<<<<<< HEAD
 
+=======
+>>>>>>> 16bb2e7... Moved /register contents into a util file (#431)
```
Comment (hsorathia) [inline api/util/registerUser.js]: oh man

File: api/util/registerUser.js [hunk 2]
```diff
@@ -38,7 +41,10 @@ async function registerUser(userToAdd){
       result.status = 'BAD_REQUEST';
       return result;
     }
+<<<<<<< HEAD
 
+=======
+>>>>>>> 16bb2e7... Moved /register contents into a util file (#431)
```
Comment (hsorathia) [inline api/util/registerUser.js]: oh no

File: api/models/User.js [hunk 3]
```diff
@@ -75,13 +75,12 @@ UserSchema.pre('save', function(next) {
...
           return next(error);
         }
-
         member.password = hash;
         return next();
       });
     });
   } else {
-    return next();
+    return next(member);
```
Comment (hsorathia) [inline api/models/User.js:83]: probably should revert this back

Comment (hsorathia) [reply api/models/User.js:83]: actually not sure, checks still pass

Comment (evanugarte) [reply api/models/User.js:83]: if it works without it then revert it

**Comment by hsorathia** [comment]
> Looks good so far, few things:
> 
> 1. Add a fullfillment message e.g. `Success! Your account has been created.`
> 2. Make sure the tool exits after the account is created instead of having it hang like this:
>    image
> 3. Add a script to our package.json in our topmost directory so we can run this tool like:
> 
> ```
> npm run create-user
> ```

should be done by now

File: api/CreateUserScript.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      filter: function(val) {
+        console.log(val);
+        return val;
+      }
+    }
+  ])
+  .then(async (answers) => {
+    const newUser = new User({
+      password: answers.password,
+      firstName: 'Dummy',
```
Comment (jerrylee17) [inline api/CreateUserScript.js:59]: prompt for this?

File: api/CreateUserScript.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      lastName: 'Account',
```
Comment (jerrylee17) [inline api/CreateUserScript.js:60]: prompt for this?

Comment (evanugarte) [reply api/CreateUserScript.js:60]: I agree

File: api/CreateUserScript.js (hunk 4 above)
Comment (evanugarte) [reply api/CreateUserScript.js:59]: I agree

File: api/CreateUserScript.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+async function mySave(userToSave) {
+  let done = false;
+  await userToSave.save()
+    .then(() => {
+      console.log('Account creation successful');
+      done = true;
+    })
+    .catch(err => {
+      console.log('Some error occurred: ', err);
```
Comment (evanugarte) [inline api/CreateUserScript.js:100]:
```suggestion
      console.log('An error occurred trying to create the account: ', err);
```

File: api/models/User.js [hunk 7]
```diff
@@ -75,7 +75,6 @@ UserSchema.pre('save', function(next) {
         if (error) {
           return next(error);
         }
```
Comment (evanugarte) [inline api/models/User.js:77]: revert the changes in this file

File: api/util/registerUser.js [hunk 8]
```diff
@@ -24,7 +24,6 @@ async function registerUser(userToAdd){
       email: userToAdd.email.toLowerCase(),
       major: userToAdd.major || ''
     });
```
Comment (evanugarte) [inline api/util/registerUser.js:26]: revert the changes in this file

File: package.json [hunk 9]
```diff
@@ -11,7 +11,8 @@
     "frontend": "react-scripts start",
     "build": "react-scripts build",
     "deploy": "git pull && npm install && pm2 stop database api app && pm2 delete database api app && pm2 start mongod --name \"database\" && npm run build && pm2 start npm --name \"app\" -- start && pm2 start npm \"app\" -- run dev",
-    "lint": "./node_modules/.bin/eslint src/ api/models/ api/mailer/ api/routes/ api/util util/ test/"
+    "lint": "./node_modules/.bin/eslint src/ api/models/ api/mailer/ api/routes/ api/util util/ test/",
+    "create-user": "node ./api/CreateUserScript.js"
```
Comment (evanugarte) [inline package.json:15]: consider adding a `create-user` script in `api/package.json`, and then in this file you can do:
```suggestion
    "create-user": "npm run create-user --prefix api"
```

**Comment by evanugarte** [comment]
Uh oh merge conflict!!!

**Comment by evanugarte** [comment]
We need to validate the input with email regex, here are some links:
1. [rejecting a pre save](https://mongoosejs.com/docs/middleware.html#error-handling)
  - Add it [here](https://github.com/SCE-Development/Core-v4/blob/c07c47f7ca565c00d6eee21f52a7b79cb487b90a/api/models/User.js#L67)
2. [email regex](https://www.google.com/search?q=email+regex+in+js)
3. Add default values to some inputs, e.g.
```
? Please enter a password (default: "sce"):
```

**Comment by hsorathia** [comment]
> We need to validate the input with email regex, here are some links:
> 
> 1. [rejecting a pre save](https://mongoosejs.com/docs/middleware.html#error-handling)
> 
> * Add it [here](https://github.com/SCE-Development/Core-v4/blob/c07c47f7ca565c00d6eee21f52a7b79cb487b90a/api/models/User.js#L67)
> 
> 1. [email regex](https://www.google.com/search?q=email+regex+in+js)
> 2. Add default values to some inputs, e.g.
> 
> ```
> ? Please enter a password (default: "sce"):
> ```

done

File: api/CreateUserScript.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        promiseLibrary: require('bluebird'),
+        useNewUrlParser: true,
+        useUnifiedTopology: true,
+        useCreateIndex: true
+      })
+      .then(() => {
+        console.log('MongoDB Connection Successful');
+      })
+      .catch(error => {
+        console.error(error);
```
Comment (evanugarte) [inline api/CreateUserScript.js:87]: Instead of printing the error, print this to the user like below:

```
We couldn't connect to MongoDB. Are you sure the backend server is running?
```

File: api/CreateUserScript.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+      });
+    const saveUserResponse = await saveUser(newUser);
```
Comment (evanugarte) [inline api/CreateUserScript.js:89]: use a `.then` and `.catch` block here. 
```js
saveUser()
  .then(_ => {
    // print that saving the user with email <email> and password <password> was successful
  })
  .catch(_ => {
    // dont print the error, just say that the save wasn't successful and that they should try again
  })
```

File: api/CreateUserScript.js [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+  })
+  .catch(error => {
+    if(error.isTTyError) {
+      console.log(error, 'prompt couldn\'t be rendered in tty environment.');
+    } else {
+      console.log(error, 'error processing response.');
+    }
+  });
+
+async function saveUser(userToSave) {
```
Comment (evanugarte) [inline api/CreateUserScript.js:105]: move this function to the top of the file. its convention to define the function before using it

File: api/CreateUserScript.js [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    {
+      type: 'input',
+      name: 'password',
+      message: 'Please enter a password (default: "sce"):'
+    },
+    {
+      type: 'list',
+      name: 'accountLevel',
+      message: 'Choose an account level from the list below: ',
+      choices: [
```
Comment (ThaiQ) [inline api/CreateUserScript.js:34]: We have all of the [enums in src](https://github.com/SCE-Development/Core-v4/blob/c07c47f7ca565c00d6eee21f52a7b79cb487b90a/src/Enums.js#L34)

Do you think you could parse that list instead? it will dynamically respond to enum changes in the future.

Comment (ThaiQ) [reply api/CreateUserScript.js:34]: thoughts @evanugarte ?

Comment (hsorathia) [reply api/CreateUserScript.js:34]: that's a good idea actually, i'll pull that in and use those values instead.

File: api/util/CreateUserScript.js [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+          value: membershipState.PENDING
+        },
+        {
+          name: 'Banned',
+          value: membershipState.BANNED
+        }
+      ],
+      filter: function(val) {
+        // eslint-disable-next-line no-console
+        console.log(val);
```
Comment (evanugarte) [inline api/util/CreateUserScript.js:70]: change every instance of console.log in this file to console.debug

the reason we do this is because:
1. we know people use console.log for printing and we want to discourage leaving print statements in code
2. for the print statements we do feel need to stay, we use console.debug to avoid limiting errors

File: api/util/CreateUserScript.js [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+                    + ((answers.password) ? answers.password : 'sce')
+                    + ' succesfully created.', defaultColor);
+        this.mongoose.connection.close();
+      })
+      .catch(_=> {
+        // eslint-disable-next-line no-console
+        console.log(redColor, 'Account creation unsuccessful. Please try'
+                                + ' running the user creation script'
+                                + ' again.', defaultColor);
+        this.mongoose.connection.close();
```
Comment (evanugarte) [inline api/util/CreateUserScript.js:113]: you need to put a process.exit in this catch block too
